### PR TITLE
Support pluggable config providers

### PR DESCRIPTION
**Pull request by feedmeapples**
## What was changed

Adds support for pluggable Config providers:

* adds Config Provider interface
* it's wrapped in a Config Provider With Refresh (that periodically gets the config and caches it)
* there is a default implementation of Config Provider that is used in our docker image (fs_config_provider)
* people can start the server with their own provider instead  

## Why?

Flexibility in how the config is provided
TLS params periodic refresh

## Checklist

1. Closes 

2. How was this tested:

* First start Temporal service with mTLS disabled
* In UI server's config set `refreshInterval: 30s`, ex in development.yaml file
* Load** UI server and open http://localhost:8080/ and verify namespaces are loaded
* Stop Temporal service
* Start a new Temporal service with mTLS enabled using https://github.com/temporalio/samples-server/tree/main/tls/tls-simple
* Go to UI server development.yaml and set values for `caFile`, `certFile`, `keyFile`, `enableHostVerification`, `serverName`
* Wait up to 30 seconds (refreshInterval) until you see `2022[/02/16]() 11:38:06 loaded new UI server configuration` in the logs, should also print that it loaded new TLS CA, KEY etc
* Open  http://localhost:8080/ again and verify that the namespaces are loaded

3. Any docs updates needed?

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -70,23 +68,28 @@ var swaggeruiAssets embed.FS
...
+		httpServer  *echo.Echo
+		options     *server_options.ServerOptions
+		cfgProvider *config.ConfigProviderWithRefresh
 	}
 )
 
 // NewServer returns a new instance of server that serves one or many services.
 func NewServer(opts ...server_options.ServerOption) *Server {
 	serverOpts := server_options.NewServerOptions(opts)
+	cfgProvider, err := config.NewConfigProviderWithRefresh(*serverOpts.ConfigProvider)
```
Comment (sergeybykov) [inline server/server.go:80]: This `err` is left unchecked, which may lead to `cfgProvider` being `nil`.

File: server/routes/api.go [hunk 2]
```diff
@@ -35,20 +36,33 @@ import (
...
-func temporalAPIHandler(temporalConn *grpc.ClientConn) echo.HandlerFunc {
+func temporalAPIHandler(cfgProvider *config.ConfigProviderWithRefresh) echo.HandlerFunc {
 	return func(c echo.Context) error {
-		mux, err := getTemporalClientMux(c, temporalConn)
+		cfg, err := cfgProvider.GetConfig()
+		if err != nil {
+			return c.JSON(http.StatusInternalServerError, err)
+		}
+
+		tls, err := rpc.CreateTLSConfig(cfg.TemporalGRPCAddress, &cfg.TLS)
```
Comment (sergeybykov) [inline server/routes/api.go:58]: Is this called on every request and never cached?

File: server/routes/auth.go [hunk 3]
```diff
@@ -52,18 +52,19 @@ type Claims struct {
 }
 
 // SetAuthRoutes sets routes used by auth
-func SetAuthRoutes(e *echo.Echo, cfg *config.Auth) {
+func SetAuthRoutes(e *echo.Echo, cfgProvider *config.ConfigProviderWithRefresh) {
 	ctx := context.Background()
+	cfg, _ := cfgProvider.GetConfig()
```
Comment (sergeybykov) [inline server/routes/auth.go:57]: Why ignore error?

File: server/routes/api.go (hunk 2 above)
Comment (feedmeapples) [reply server/routes/api.go:58]: it's called on every request though receives a cached config https://github.com/temporalio/ui-server/pull/102/files#diff-0e504256c68c0d5b7cd3c799aa6c942469504ebb007e337f33cef429241061a1R67

Comment (sergeybykov) [reply server/routes/api.go:58]: Config passed into `rpc.CreateTLSConfig()` is cashed, but the output of it is generated every time? And then `rpc.CreateGRPCConnection()` is also called every time. Shouldn't they be cached?

Comment (feedmeapples) [reply server/routes/api.go:58]: right, figured to cached the GRPC connections once we add multi-cluster support and work with the pool of connections. Can add caching now

**Review by stevekinney — APPROVED** [review]

**Review by sergeybykov — APPROVED** [review]

**Pull request merged**